src/ray/gcs/pubsub/gcs_pub_sub.cc

```diff
@@ -314,6 +406,36 @@ Status GcsSubscriber::SubscribeAllNodeInfo(
   return pubsub_->SubscribeAll(NODE_CHANNEL, on_subscribe, done);
 }
 
+Status GcsSubscriber::SubscribeAllNodeResources(
+    const ItemCallback<rpc::NodeResourceChange> &subscribe, const StatusCallback &done) {
+  if (subscriber_ != nullptr) {
+    auto subscribe_item_callback = [subscribe](const rpc::PubMessage &msg) {
+      RAY_CHECK(msg.channel_type() == rpc::ChannelType::GCS_NODE_RESOURCE_CHANNEL);
+      subscribe(msg.node_resource_message());
+    };
+    auto subscription_failure_callback = [](const std::string &, const Status &status) {
+      RAY_LOG(WARNING) << "Subscription to NodeResource channel failed: "
+                       << status.ToString();
+    };
+    if (subscriber_->SubscribeAll(
+            std::make_unique<rpc::SubMessage>(),
+            rpc::ChannelType::GCS_NODE_RESOURCE_CHANNEL, gcs_address_,
+            [done](Status status) { done(status); }, std::move(subscribe_item_callback),
+            std::move(subscription_failure_callback))) {
+      return Status::OK();
+    }
+    return Status::ObjectExists(
+        "NodeResource channel already subscribed. Please unsubscribe first if it needs "
+        "to be resubscribed.");
+  }
+  auto on_subscribe = [subscribe](const std::string &, const std::string &data) {
```

Comment:
Please use else here to make it clearer.

Reply:
Using an `else` block would add unnecessary indentation which is less readable IMO. In fact I have enabled [this clang-tidy check](https://clang.llvm.org/extra/clang-tidy/checks/readability-else-after-return.html) for awhile ..

Reply:
When I'm reading this, I'm actually lost since I'm not sure whether 431 will be executed or not. But with if-else, it should be clearer. I'm feeling it makes it easier to read and less prone to make mistake.

Reply:
Chatted off thread. Added comment to indicate the separate control flows for GCS / Redis subscribers. In future the Redis subscriber will be removed.